src/ray/common/asio/io_context.cc

```diff
@@ -0,0 +1,257 @@
+        }
+        // Update global stats.
+        {
+          absl::WriterMutexLock lock(&(global_stats_.mutex));
+          // Global queue stats.
+          global_stats_.stats.cum_queue_time += queue_time_ns;
+          if (global_stats_.stats.min_queue_time > queue_time_ns) {
+            global_stats_.stats.min_queue_time = queue_time_ns;
+          }
+          if (global_stats_.stats.max_queue_time < queue_time_ns) {
+            global_stats_.stats.max_queue_time = queue_time_ns;
+          }
+          // Global execution stats.
+          global_stats_.stats.cum_execution_time += execution_time_ns;
+          if (global_stats_.stats.min_execution_time > execution_time_ns) {
+            global_stats_.stats.min_execution_time = execution_time_ns;
+          }
+          if (global_stats_.stats.max_execution_time < execution_time_ns) {
+            global_stats_.stats.max_execution_time = execution_time_ns;
+          }
+          // Global current count.
+          global_stats_.stats.curr_count--;
+        }
+      });
+  boost::asio::io_context::post(std::move(wrapped_handler));
+}
+
+/// A helper for creating a snapshot view of the stats for a handler.
+/// This acquires a reader lock on the provided guarded handler stats, and creates a
+/// lockless copy of the stats.
+HandlerStats to_stats_view(const GuardedHandlerStats &stats) {
+  absl::ReaderMutexLock lock(&(stats.mutex));
+  return HandlerStats(stats.stats);
+}
+
+HandlerStats io_context_proxy::get_global_stats() const {
+  return to_stats_view(global_stats_);
+}
+
+absl::optional<HandlerStats> io_context_proxy::get_handler_stats(
+    const std::string &handler_name) const {
+  absl::ReaderMutexLock lock(&mutex_);
+  auto it = post_handler_stats_.find(handler_name);
+  if (it == post_handler_stats_.end()) {
+    return {};
+  }
+  return to_stats_view(*it->second);
+}
+
+std::vector<std::pair<std::string, HandlerStats>> io_context_proxy::get_handler_stats()
+    const {
+  // We lock the stats table while copying the table into a vector.
+  absl::ReaderMutexLock lock(&mutex_);
+  std::vector<std::pair<std::string, HandlerStats>> stats;
+  stats.reserve(post_handler_stats_.size());
+  std::transform(
+      post_handler_stats_.begin(), post_handler_stats_.end(), std::back_inserter(stats),
+      [](const std::pair<std::string, std::shared_ptr<GuardedHandlerStats>> &p) {
+        return std::make_pair(p.first, to_stats_view(*p.second));
+      });
+  return stats;
+}
+
+std::string to_human_readable(double duration) {
+  static const std::array<std::string, 4> to_unit{"ns", "us", "ms", "s"};
+  size_t idx = std::min(to_unit.size() - 1,
+                        static_cast<size_t>(std::log(duration) / std::log(1000)));
+  double new_duration = duration / std::pow(1000, idx);
+  std::stringstream result;
+  result << std::fixed << std::setprecision(3) << new_duration << " " << to_unit[idx];
+  return result.str();
+}
+
+std::string to_human_readable(int64_t duration) {
```

Comment:
Why don't we make this function inlined?

Reply:
Hmm I thought that most compilers would inline this and the other function implicitly, but I don't see the harm in doing it explicitly. I'll add the specifier.